### ISSUE TITLE
The combined solver, incremental solving, and activation literals

### ISSUE THREAD

**Issue by martin-neuhaeusser** [Z3Prover/z3#501]
Our application uses Z3 and activation literals for incremental solving quite heavily.
We use a configuration with a combined solver, where the incremental solver (which lacks sophisticated preprocessing in its tactics) has a rather short timeout, and the non-incremental solver uses a much longer timeout.

We introduce activation literals not by using assert_and_track, but by asserting implications of the form
a_i => formula_i on the solver stack. Then check_sat is called incrementally with subsets of these a_i.

Against our expectations, we are ending up using the incremental solver without the desired short timeout being imposed by the combined_solver implementation. The fallback to the non-incremental solver does not occur. 

The reason seems to be that the code in https://github.com/Z3Prover/z3/blob/71fff8ffa2fa80a34d2b3bf3f705ef0852310b04/src/solver/combined_solver.cpp#L196-L202 shortcuts the entire combined_solver heuristic and simply applies the (often slow) incremental solver2.
I wonder why this is necessary. As far as I understood the Z3 sources, activation literals that are defined by asserting implications as sketched above, are held in sync between solver1 and solver2. 
So it should be possible to use the combined_solver heuristic for such queries, too.
The only thing one has to change is to propagate num_assumptions and assumptions into the check_sat calls for solver1 and solver2. Some experiments that I have run seem to support that claim.

Is there a good reason for always resorting to the incremental solver once activation literals are specified? Or is this only necessary for activation literals created via the assert_and_track method?
Could the combined_solver distinguish between these two scenarios?
This might also be related commit 9b137d54d34e0a014b4c5c5c9c21cbf43b78f604

**Comment by Adorf** [comment]
We are also interested in incremental solving. Where can I learn something about it?

**Comment by NikolajBjorner** [comment]
> Is there a good reason for always resorting to the incremental solver once activation literals are specified?

I take it is a matter of what was possible to test and get working.

Commit  9b137d5 indicates that not all tactics support cores and the previous accounting for assumptions was inaccurate and therefore had to disable this pass for tracked assumptions.
With some extra time we can double check what happens if we disable the check for assumptions on the example scenario when assertions are tracked.

There is an alternative available over the API: you create a tactic corresponding to the 
class of formulas you are solving, and then create a solver based on that tactic. 
This would correspond to the non-incremental solver. 
Over the SMT-LIB interface it is (check-sat-using ...).

**Comment by martin-neuhaeusser** [comment]
I did not experiment at all with assumptions that have been introduced by assert_and_track. But for assertions of the form a_i => formula_i, it seems to work (well, at least in our benchmarks) as expected.
The change that I made for my tests is available here: https://github.com/martin-neuhaeusser/z3/commit/45aaed6d96df1cdb2a504ec25c0d59685ed9450c

Is using the API to construct a tactic really an alternative? Can such a solver be combined with an incremental solver, easily?

Our original idea (which is, admittedly, based on an "uneducated guess") was to let the incremental solver try to solve the problem first (but only for a short amount of time) and if it fails to produce an answer, to automatically resort to a non-incremental solver. In this way, we hoped to solve simple instances incrementally whereas leaving the hard instances to a more powerful solver. We have used this in a predicate abstraction framework, where predicates are represented by activation literals. So our expectation was that an incremental solver has the chance to learn facts about predicate formulas that it can re-use in subsequent queries that involve the same predicates.

I hope to be able to produce a benchmark to see if the change has any impact on our verifier's performance.

**Comment by NikolajBjorner** [comment]
Was there an impact?

**Comment by martin-neuhaeusser** [comment]
When I checked in March, there was no measurable impact of the patch; 

I just ran a subset of our benchmarks with three different engines again with two versions of Z3: Once with commit 56c7875 and once with 56c7875 extended with the patch.
As a result, we observe a notable benefit (895s vs. 701s) of the patch only on one benchmark (out of 37) and only with our TreeIC3 engine. For all other instances, there was no impact at all.

That there is hardly any effect puzzles me a bit, given that I would have expected the two solvers managed by the combined solver to exhibit quite different performance characteristics. I will look into this a bit more closely, once I have time. And run a larger benchmark suite.

Anyways, the good news is that no errors occured in our benchmarks if the non-incremental solver is used with activation literals (i.e. when activation literals are encoded as implications in the assertions).
